src/formats/SMI.cpp

```diff
@@ -441,13 +394,88 @@ static void find_rings(
     while (!all(hit_atoms)) {
         auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
         auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
-        auto& current_atom_bonds = adj_list[current_atom];
+
+        // mark this atom as processed
         hit_atoms[current_atom] = true;
-        if (current_atom_bonds.empty()) {
+
+        // it has no connections... therefore no rings!
+        if (adj_list[current_atom].empty()) {
             continue;
         }
 
-        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms, current_atom_bonds[0], current_atom);
+        // Search to find all rings for the component (all connected atoms)
+        // for the 'current' atom.
+        // Not guaranteed to fast, nor efficient, but just to work
+        // Also not guaranteed to find the SSSR, but it will find all rings
+        // needed for a given structure to be writen as SMILES
+        // adj_list The adjacentcy list graph
+        // hit_atoms Atoms already encountered
+        // ring_bonds Bonds already known to ring forming bonds
+        // ring_atoms Map from an atom to the number of rings it forms
+
+        // The atoms to process consist of :
+        // current_atom Atom to be checked in this iteration
+        // previous_atom Atom which was bound to the current atom
+
+        std::stack<std::pair<size_t, size_t>> atoms_to_process;
+        atoms_to_process.push({ current_atom, current_atom });
+
+        while (atoms_to_process.size()) {
+
+            size_t previous_atom;
+
+            std::tie(current_atom, previous_atom) = atoms_to_process.top();
+            atoms_to_process.pop();
+
+            auto& current_atom_bonds = adj_list[current_atom];
+
+            // We go through the neighbors backwards because we are using a
+            // stack instead and want to through them in the forward direction
```

Comment:
I am not sure I understand this. Is a word missing after **instead**?

Reply:
It's meant to refer to later in the paragraph. Removing the word actually just makes it easier to read